src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -205,36 +239,81 @@ uct_cuda_ipc_open_memhandle_mempool(uct_cuda_ipc_rkey_t *key,
                                     CUdeviceptr *mapped_addr)
 {
     CUmemAccessDesc access_desc = {};
-    CUdeviceptr dptr;
-    ucs_status_t status;
+    ucs_status_t status         = UCS_OK;
+    char *hkey;
+    CUmemoryPool *mpool;
+    khiter_t khiter;
+    int khret;
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportFromShareableHandle(
-                &key->ph.pool, (void *)&key->ph.handle.fabric_handle,
-                CU_MEM_HANDLE_TYPE_FABRIC, 0));
-    if (status != UCS_OK) {
-        return status;
-    }
+    pthread_rwlock_rdlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    khiter = kh_get(cuda_ipc_rem_mpool_cache,
+                    &uct_cuda_ipc_rem_mpool_cache.hash,
+                    (char*)&key->ph.handle.fabric_handle);
+    if (ucs_likely(khiter != kh_end(&uct_cuda_ipc_rem_mpool_cache.hash))) {
+        key->ph.pool = kh_val(&uct_cuda_ipc_rem_mpool_cache.hash, khiter);
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(&dptr,
-                key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
-    if (status != UCS_OK) {
-        return status;
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                    *key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+        goto err;
     }
 
-    status = uct_cuda_ipc_init_access_desc(&access_desc);
-    if (status != UCS_OK) {
-        return status;
+    /* acquire writer lock as we did not find cached mpool in hashmap */
+    pthread_rwlock_unlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    pthread_rwlock_wrlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+
+    hkey   = ucs_strndup((char*)&key->ph.handle.fabric_handle,
+                         sizeof(CUmemFabricHandle), "fabric_handle");
```

Comment:
then we also need to check if ucs_strndup returned NULL..
Better avoid that with struct I described above

Reply:
Hi @iyastreb, 
This is still not the common case and the overhead of allocating the key occurs whenever key was found in the hashmap. If I understand correctly, your suggestion is to avoid the call to `ucs_strndup` and have a local variable (of type `cuda_ipc_rem_mpool_cache_key_t`) in the function to avoid heap allocation overhead. However, the reason why have to duplicate the fabric handle (`key->ph.handle.fabric_handle`) and not just use it is because it appears that the hashmap expects the keys to have the same lifetime as the hashmap (maybe because the hashmap data structure doesn't make a copy of the key). As the fabric handle lifetime expires after the associated rkey's role completes, using a copy of the fabric handle allows the keys to persist until a later point when we destroy both the keys and the hashmap itself. For this reason, I think dynamic allocation of they hash key is unavoidable. Let me know if I've mistaken your suggestion. For now, I'll only add logic to handle ucs_strndup returning NULL.

Reply:
Actually khash does copy keys, and key can be anything: number, pointer, structure, whatever:
```C
#define __KHASH_TYPE(name, khkey_t, khval_t) \
	typedef struct kh_##name##_s { \
		khkey_t *keys; \

	SCOPE khint_t kh_put_##name(kh_##name##_t *h, khkey_t key, int *ret) \
	{																	\
...
			h->keys[x] = key;											\
```
So it can perfectly work with structures as well. I admit that there could be some consequences of having structure as a key:
 - khash preallocates key array, which may provoke higher memory usage.
 But in this case I think it's negligible, because even on maximum size of 100 with dynamic allocation we store:
 128x8B (pointer key) + 100x64B (dynamic key) = 7424B
 128x64B (fixed struct key) = 8192B
 - key is passed **by value** to hash_code and equals functions
This is optimised away by compiler (I checked that), and code without pointer actually works faster.

Let me do this little change along with some other minor fixes

Reply:
> Actually khash does copy keys, and key can be anything: number, pointer, structure, whatever:

@iyastreb Thanks for the explanation. If khash does perform a copy, do we need an explicit structure(`cuda_ipc_rem_mpool_cache_key_t`) for the key? Why not just use `CUmemFabricHandle`? Also, if copies are being performed, was it the case that previously pointers to the key were being copied as opposed to the key contents which caused the failures that I saw?

Reply:
But this is exactly what I proposed previously: https://github.com/openucx/ucx/pull/10082#discussion_r1743138544
You said "This was previously the case but it ran into cache lookup issues when the actual key was freed". But it seems that previously you tried to store the key by **pointer**, not by value.
Of course I agree that it's better to keep things simple and just use `CUmemFabricHandle` as a key.
I implemented this change